cephfs/statx.go

```diff
@@ -0,0 +1,133 @@
+package cephfs
+
+/*
+#cgo LDFLAGS: -lcephfs
+#cgo CPPFLAGS: -D_FILE_OFFSET_BITS=64
+#include <cephfs/libcephfs.h>
+*/
+import "C"
+
+// StatxWant values contain bit-flags indicating what data should be
+// populated by a statx-type call.
+type StatxWant uint32
+
+const (
+	// StatxMode requests the mode value be filled in.
+	StatxMode = StatxWant(C.CEPH_STATX_MODE)
+	// StatxNlink requests the nlink value be filled in.
+	StatxNlink = StatxWant(C.CEPH_STATX_NLINK)
+	// StatxUid requests the uid value be filled in.
+	StatxUid = StatxWant(C.CEPH_STATX_UID)
+	// StatxRdev requests the rdev value be filled in.
+	StatxRdev = StatxWant(C.CEPH_STATX_RDEV)
+	// StatxAtime requests the access-time value be filled in.
+	StatxAtime = StatxWant(C.CEPH_STATX_ATIME)
+	// StatxMtime requests the modified-time value be filled in.
+	StatxMtime = StatxWant(C.CEPH_STATX_MTIME)
+	// StatxIno requests the inode be filled in.
+	StatxIno = StatxWant(C.CEPH_STATX_INO)
+	// StatxSize requests the size value be filled in.
+	StatxSize = StatxWant(C.CEPH_STATX_SIZE)
+	// StatxBlocks requests the blocks value be filled in.
+	StatxBlocks = StatxWant(C.CEPH_STATX_BLOCKS)
+	// StatxBasicStats requests all the fields that are part of a
+	// traditional stat call.
+	StatxBasicStats = StatxWant(C.CEPH_STATX_BASIC_STATS)
+	// StatxBtime requests the birth-time value be filled in.
+	StatxBtime = StatxWant(C.CEPH_STATX_BTIME)
+	// StatxVersion requests the version value be filled in.
+	StatxVersion = StatxWant(C.CEPH_STATX_VERSION)
+	// StatxAllStats requests all known stat values be filled in.
+	StatxAllStats = StatxWant(C.CEPH_STATX_ALL_STATS)
+)
+
+// AtFlags represent flags to be passed to calls that control how files
+// are used or referenced. For example, not following symlinks.
+type AtFlags uint
+
+const (
+	// AtNoAttrSync requests that the stat call only fetch locally-cached
+	// values if possible, avoiding round trips to a back-end server.
+	AtNoAttrSync = AtFlags(C.AT_NO_ATTR_SYNC)
+	// AtSymlinkNofollow indicates the call should not follow symlinks
+	// but operate on the symlink itself.
+	AtSymlinkNofollow = AtFlags(C.AT_SYMLINK_NOFOLLOW)
+)
+
+// CephStatx instances are returned by extended stat (statx) calls.
+// Note that CephStatx results are similar to but not identical
+// to (Linux) system statx results.
+type CephStatx struct {
+	statx C.struct_ceph_statx
+}
+
+// Mask returns a value containing the bits indicating what
+// statx fields have been set.
+func (csx *CephStatx) Mask() StatxWant {
+	return StatxWant(csx.statx.stx_mask)
+}
+
+// Blksize returns the block size.
+func (csx *CephStatx) Blksize() uint32 {
```

Comment:
I wonder if these functions should return `(value, error)` and return an error in case the `stx_mask` does not contain the bit for the requested field. `Mask()` would not return an error, as that should always be set.

Or, do not provide accessors but set `CephStatx.Blksize` as a public member.

Reply:
The reason I went with methods are:
* I didn't want to duplicate the fields of the struct ceph_statx because we were not going to drop many fields
* To more closely match the approach in directory.go with DirEntry
* To allow for calling libs to use an interface for CephStatx results

I think that the 99% use case is that cephfs will return the content requested and returning an error in addition to the value is more complexity than needed. I'd rather add methods like `HasMask(*CephStatx, StatxMax) bool` to aid the user wrt to checking the bit fields.

Reply:
By using accessors, I expect some validation (either returning an error or boolean). A `HasMask()` function as proposed below can be an alternative is you are really fixed on the `DirEntry` match. That is fine with me too. Offering ways to not use bitmasks and checking has my preference.

However I see a `Stat()` and `Statx()` as quite different things. `Readdir()` and `Access()` (based on `DirEntry`) are more similar, same as `ReaddirPlus()` and `Stat()`. `Statx()` is an even more extended version with optional members (with `CephStatx` maybe more like a `map`?).